File: frontend/src/types/certification-record.ts
export type CertificationScheme =
  | "eu_cybersecurity_act"
  | "iec_62443"
  | "common_criteria"
  | "etsi_en_303_645"
  | "iso_iec_27001"
  | "soc2"
  | "other";

export type CertificationStatus =
  | "pending"
  | "active"
  | "expired"
  | "suspended"
  | "withdrawn";

export interface CertificationRecordArtifactLink {
  id: string;
  artifact_revision: {
    id: string;
    artifact_id: string;
    revision_number: number;
    source_type: 'upload' | 'external_link';
    original_filename: string | null;
    content_type: string | null;
    file_size_bytes: number | null;
    sha256: string | null;
    storage_path: string | null;
    external_url: string | null;
    change_summary: string | null;
    uploaded_by_user_id: string;
    created_at: string;
  };
  linked_by_user_id: string | null;
}

export interface CertificationRecord {
  id: string;
  product_id: string;
  certification_scheme: CertificationScheme;
  certification_scheme_label: string | null;
  certification_body_name: string;
  certificate_number: string | null;
  scope_description: string;
  issued_date: string | null;
  valid_until_date: string | null;
  status: CertificationStatus;
  notes: string | null;
  recertification_required_by: string | null;
  artifact_links?: CertificationRecordArtifactLink[] | null;
  created_at: string;
  updated_at: string;
}

export interface CertificationRecordCreate {
  product_id: string;
  certification_scheme: CertificationScheme;
  certification_scheme_label?: string | null;
  certification_body_name: string;
  certificate_number?: string | null;
  scope_description: string;
  issued_date?: string | null;
  valid_until_date?: string | null;
  status?: CertificationStatus;
  notes?: string | null;
  recertification_required_by?: string | null;
}

export interface CertificationRecordUpdate {
  certification_scheme?: CertificationScheme;
  certification_scheme_label?: string | null;
  certification_body_name?: string;
  certificate_number?: string | null;
  scope_description?: string;
  issued_date?: string | null;
  valid_until_date?: string | null;
  status?: CertificationStatus;
  notes?: string | null;
  recertification_required_by?: string | null;
}

export const SCHEME_LABELS: Record<CertificationScheme, string> = {
  eu_cybersecurity_act: "EU Cybersecurity Act (EUCC)",
  iec_62443: "IEC 62443",
  common_criteria: "Common Criteria (ISO 15408)",
  etsi_en_303_645: "ETSI EN 303 645",
  iso_iec_27001: "ISO/IEC 27001",
  soc2: "SOC 2",
  other: "Other",
};

export const STATUS_LABELS: Record<CertificationStatus, string> = {
  pending: "Pending",
  active: "Active",
  expired: "Expired",
  suspended: "Suspended",
  withdrawn: "Withdrawn",
};

File: backend/app/schemas/certification_record.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.models.enums import CertificationScheme, CertificationStatus
from app.schemas.common import TimestampedRead


class CertificationRecordCreate(BaseModel):
    product_id: UUID
    certification_scheme: CertificationScheme
    certification_scheme_label: str | None = Field(default=None, max_length=255)
    certification_body_name: str = Field(min_length=1, max_length=255)
    certificate_number: str | None = Field(default=None, max_length=255)
    scope_description: str = Field(min_length=1)
    issued_date: date | None = None
    valid_until_date: date | None = None
    status: CertificationStatus = CertificationStatus.pending
    notes: str | None = None
    recertification_required_by: date | None = None

    @model_validator(mode="after")
    def validate_dates_and_label(self) -> "CertificationRecordCreate":
        if self.issued_date and self.valid_until_date and self.valid_until_date < self.issued_date:
            raise ValueError("valid_until_date must be on or after issued_date")
        if self.certification_scheme.value != "other":
            self.certification_scheme_label = None
        return self


class CertificationRecordUpdate(BaseModel):
    certification_scheme: CertificationScheme | None = None
    certification_scheme_label: str | None = Field(default=None, max_length=255)
    certification_body_name: str | None = Field(default=None, min_length=1, max_length=255)
    certificate_number: str | None = Field(default=None, max_length=255)
    scope_description: str | None = Field(default=None, min_length=1)
    issued_date: date | None = None
    valid_until_date: date | None = None
    status: CertificationStatus | None = None
    notes: str | None = None
    recertification_required_by: date | None = None

    @model_validator(mode="after")
    def validate_dates_and_label(self) -> "CertificationRecordUpdate":
        if self.issued_date and self.valid_until_date and self.valid_until_date < self.issued_date:
            raise ValueError("valid_until_date must be on or after issued_date")
        if self.certification_scheme and self.certification_scheme.value != "other":
            self.certification_scheme_label = None
        return self


class CertificationRecordArtifactLinkRead(BaseModel):
    id: UUID
    artifact_revision: Any
    linked_by_user_id: UUID | None


class CertificationRecordRead(TimestampedRead):
    product_id: UUID
    certification_scheme: CertificationScheme
    certification_scheme_label: str | None
    certification_body_name: str
    certificate_number: str | None
    scope_description: str
    issued_date: date | None
    valid_until_date: date | None
    status: CertificationStatus
    notes: str | None
    recertification_required_by: date | None
    artifact_links: list[CertificationRecordArtifactLinkRead] | None = None
